src/main/java/org/opentripplanner/routing/core/StateEditor.java

```diff
@@ -36,48 +36,53 @@ public StateEditor(State parent, Edge e) {
     child = parent.clone();
     child.backState = parent;
     child.backEdge = e;
-    // We clear child.next here, since it could have already been set in the
-    // parent
+    // We clear child.next here, since it could have already been set in the parent
     child.next = null;
+
+    final Vertex parentVertex = parent.vertex;
+
     if (e == null) {
       child.backState = null;
-      child.vertex = parent.vertex;
+      child.vertex = parentVertex;
       child.stateData = child.stateData.clone();
-    } else if (e.getFromVertex() == null || e.getToVertex() == null) {
-      child.vertex = parent.vertex;
+      return;
+    }
+
+    final Vertex fromVertex = e.getFromVertex();
+    final Vertex toVertex = e.getToVertex();
+
+    if (fromVertex == null || toVertex == null) {
+      child.vertex = parentVertex;
       child.stateData = child.stateData.clone();
       LOG.error("From or to vertex is null for {}", e);
       defectiveTraversal = true;
+      return;
+    }
+
+    // Note that we use equals(), not ==, here to allow for dynamically created vertices
+    if (parentVertex.equals(fromVertex)) {
```

Comment:
I think I understand why you did that but perhaps you can confirm this in the dev meeting.